Feat: Clumsy table climbs

Vladisvell opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
<!-- Опишите, что делает ваш ПР. Документировать каждую деталь не требуется, просто укажите основные изменения. -->

Обновление, добавляющее опрокидывание предметов со стола в попытках на него залезть, потанцевать или кинуть на него кого-либо.

## Ссылка на предложение/Причина создания ПР
<!-- Здесь оставьте ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что ваше предложение одобрено. Либо укажите, почему этот ПР должен пройти без предложки. -->
<!-- Пример ссылки: https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->

https://discord.com/channels/617003227182792704/755125334097133628/1087419249917702264
https://discord.com/channels/617003227182792704/755125334097133628/1087419262454472775

## Демонстрация изменений
<!-- Здесь вы можете показать изменения внешне, к примеру новые спрайты, звуки или изменения карты. Или написать, что именно изменилось для игроков. Этот пункт полностью опционален и его можно удалить. -->
![rolling stones2](https://user-images.githubusercontent.com/73733747/226257368-7fd93fc8-1855-4167-8714-0f5e9e327ccb.gif)
![dance dance](https://user-images.githubusercontent.com/73733747/226257252-1022879e-0fca-44b9-9a0e-eefed52f3d57.gif)
![Bowling_2](https://user-images.githubusercontent.com/73733747/226257186-aacf182c-9ce3-4aa2-a397-790d7ff6a426.gif)

Vladisvell committed:
Первый коммит, добавлена функция clumse_stuff для раскидывания предметов с тайла на соседние.
Используется в do_climb

Vladisvell committed:
Merge branch 'ss220-space:master220' into sloppy-table-climbs

Vladisvell committed:
Теперь сила, с которой отлетают предметы, уменьшается в 10 раз, чтобы предотвратить абуз осколками.
-Танцы на столе разбрасывают предметы.
-Кидание персонажей на стол также разбрасывают предметы.

Vladisvell committed:
Убран лишний комментарий

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,54 @@
+			if(MOB_SIZE_SMALL) slopchance = 20
+			if(MOB_SIZE_TINY) slopchance = 10
+
+	slopchance = clamp(slopchance,0,100)
```
```suggestion
	slopchance = clamp(slopchance, 0, 100)
```

Vladisvell replied:
сделано

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,54 @@
+		switch(user.mob_size)
+			if(MOB_SIZE_LARGE) slopchance = 100
+			if(MOB_SIZE_SMALL) slopchance = 20
+			if(MOB_SIZE_TINY) slopchance = 10
```
Может лучше модификаторы сделать?
Сначала смотрим размер. От него определяем текущий slopchance. 10, 20, 80, 100.
После чего "ishuman(user)" с проверками на ДНК и Мимность, где будет +20 или -30 поверх этого шанса.
Ну и далее если оно "вдруг" перейдет границы, как я и говорил, кламп его уравняет.

Тогда и потенциально в будущем будет возможно модицифировать и дальше прок. Под те же расы (разумеется модификации не будет. Но так будет структурно лучше)

Vladisvell replied:
Изменено в следующем коммите

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,54 @@
+			if(!(AM == user || AM.anchored || isliving(AM)))
+				thrownatoms += AM
+			if(thrownatoms.len >= max_throws_count)
+				break
```
Не согласен немного с таким решением, но допустим.

Твоим решением выходит что он выбирает 15 предметов и после чего шансы на их использование. Из-за чего итого предметов откинутых может быть всего 5 штук, когда стол завален 100 штук предметов.

Я бы рекомендовал сделать счетчик, который будет проверять шанс сразу в этом проке, и если он прошел, то добавлять его в список. И когда в списке их будет 15 штук - прерывать его. Тогда не смотря на 100 штук предметов, у нас будет повышенный шанс отбросить сразу много предметов в хламе, а не 1 случайный, и при этом сохранится оптимизация.

И уже после ко всем атомам применять откидывания.

Кароче. Вынеси `if(prob(slopchance))` из 123-й строки в эту выборку в ~117 строку.

Vladisvell replied:
Сделано

PhantornRU commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -647,6 +647,10 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
 		newtonian_move(get_dir(target, src))
 		thrown_thing.throw_at(target, thrown_thing.throw_range, thrown_thing.throw_speed, src, null, null, null, move_force)
+		if(iscarbon(thrown_thing))
+			for(var/obj/structure/table in range(0, target))
```
а почему не locate() для обнаружения стола на тюрфе под целью? Я ж скидывал пример

Vladisvell replied:
Сделано

PhantornRU commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -647,6 +647,10 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
 		thrown_thing.throw_at(target, thrown_thing.throw_range, thrown_thing.throw_speed, src, null, null, null, move_force)
+		if(iscarbon(thrown_thing))
+			for(var/obj/structure/table in range(0, target))
+				var/mob/living/carbon/victim = thrown_thing
```
У нас метают не только карбонов, но и животных.
Странно будет если брошенная свинья аккуратно встанет на стол.

Vladisvell replied:
Поменял на isliving

PhantornRU commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -433,6 +433,8 @@
 				message = "радостно танцу[pluralize_ru(src.gender,"ет","ют")]!"
 				m_type = 1
 				var/dance_time = 3 SECONDS
+				for(var/obj/structure/table in range(0))
```
locate?

PhantornRU replied:
и кстати. У тебя точно выделяется table? Ибо ты задаешь что это переменная var, но не задаешь как эта переменная используется. Вот поэтому я и говорил тебе про locate.
А то так получается что ты работаешь ВООБЩЕ СО ВСЕМИ СТРУКТУРАМИ.
Т.е. я думаю если ты будешь находиться на какой-нибудь статуе клоуна, то при танцах тоже всё разметаешь в сторону.

Vladisvell replied:
Сделано

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,54 @@
 	if(get_turf(user) == get_turf(src))
 		usr.visible_message("<span class='warning'>[user] climbs onto \the [src]!</span>")
 
+
```
```suggestion
```
лишние пустые строки

Vladisvell replied:
Сделано

PhantornRU reviewed (changes requested):
Мне нравится, но нужно поправить некоторые моменты.

Vladisvell committed:
Исправления по ревью PhantomRU

Перенесена логика шансов в подбор предметов для большей рандомности (раньше бралось первые несколько подходящих, теперь рандом)
Изменена логика определения стола на тайле (теперь через Locate)
Система подсчитывания шанса опрокинуть предмет теперь работает через модификаторы
Удалены/добавлены пробелы для лучшей читаемости.

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,53 @@
+		if(man.mind.miming)
+			slopchance -= 30
+
+	slopchance = clamp(slopchance, 0, 100)
```
Маленький шанс все же должен остаться. А то он просто так переберет 100 условных предметов и ни один из них не пройдет.
```suggestion
	slopchance = clamp(slopchance, 1, 100)
```

Vladisvell replied:
йеп

PhantornRU commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,53 @@
+
+	if(ishuman(user) && user.mind)
+		var/mob/living/carbon/human/man
+		man = user
```
Выше лучше замени на данную конструкцию. А то лишняя строка на тратится буквально ни на что:
`		var/mob/living/carbon/human/man = user`

Vladisvell replied:
Готово

PhantornRU reviewed (approved):
Ну сё, в продакшн

Vladisvell committed:
Сделал код покрасивше и дал всегда ненулевой шанс на скидывание предмета

Vladisvell committed:
Убран initial, т.к ломает предметы, у которых значение force может изменяться, заменено умножением и делением на константу force_mult.

lebedev commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,53 @@
+			if(!(AM == user || AM.anchored || isliving(AM)))
+				if(prob(slopchance))
+					thrownatoms += AM
+				if(thrownatoms.len >= max_throws_count)
+					break
```
Чтоб считалось реже.
```suggestion
					if(thrownatoms.len >= max_throws_count)
						break
```

lebedev commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,53 @@
+
+	for(var/turf/T in range(0, src)) //Preventing from rotating stuff in an inventory
+		for(var/atom/movable/AM in T)
+			if(!(AM == user || AM.anchored || isliving(AM)))
```
Так как юзер живой, можно упростить проверку. И раскрыть `!`
```suggestion
			if(!AM.anchored && !isliving(AM))
```

Vladisvell replied:
Сделал

lebedev commented on code/game/objects/structures.dm:
```diff
@@ -82,9 +82,53 @@
+		if(MOB_SIZE_SMALL) slopchance = 20
+		if(MOB_SIZE_TINY) slopchance = 10
+
+	if(ishuman(user) && user.mind)
+		var/mob/living/carbon/human/man = user
+		if(/datum/dna/gene/disability/clumsy in man.active_genes)
+			slopchance += 20
+		if(man.mind.miming)
+			slopchance -= 30
```
Мимование - свойство разума, а ген неуклюжести - свойство куклы. Бездушный труп клоуна, кинутый на стол, должен все равно все раскидывать. Плюс `active_genes` есть у `/mob`, проверка на гуманность избыточна.
```suggestion
	if(/datum/dna/gene/disability/clumsy in user.active_genes)
		slopchance += 20
	if(user.mind?.miming)
		slopchance -= 30
```

Vladisvell replied:
Сделал

lebedev commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -647,6 +647,11 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
 		newtonian_move(get_dir(target, src))
 		thrown_thing.throw_at(target, thrown_thing.throw_range, thrown_thing.throw_speed, src, null, null, null, move_force)
+		if(isliving(thrown_thing))
+			var/obj/structure/table = locate(/obj/structure/table) in target.loc
```
Если локейтится такой же тип, который у переменной, то можно упростить
```suggestion
			var/obj/structure/table = locate() in target.loc
```

Vladisvell replied:
Сделано

Vladisvell committed:
Исправления по ревью lebedev
-Упрощено условие для отбора объектов для броска
-Перемещена проверка на число объектов в списке на раскидывание
-Сокращен код в поисках стола. (carbon.dm, emote.dm)

PhantornRU commented:
Авоуты пройдены, ждем залива. PepeChill
